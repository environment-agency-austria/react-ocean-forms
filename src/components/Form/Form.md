### Examples

#### Simple form

Showcase of a simple form

```jsx
import { Form, Input } from 'react-ocean-forms';

function Example() {
  // Submit callback, here you'd make your api calls
  const handleSubmit = values => {
    console.log('onSubmit, values: ' + JSON.stringify(values));
  };

  // Reset callback, can be useful in some cases for cleanup
  const handleReset = () => {
    console.log('onReset');
  };

  return (
    <Form className="demo" onSubmit={handleSubmit} onReset={handleReset}>
      <Input name="myInput" label="Example input" />
      <button type="submit">Submit</button>
      <button type="reset">Reset</button>
    </Form>
  );
}

<Example />;
```

#### Default values

Provide the form with default values that should be displayed on load. Note how the default value won't update the Field, if the user changed the input. However, on form reset the default value will be used again.

```jsx
import React, { useState } from 'react';
import { Form, Input } from 'react-ocean-forms';

function Example() {
  const [defaultValues, setDefaultValues] = useState({ myInput: 'default value' });

  // Submit callback, here you'd make your api calls
  const handleSubmit = values => {
    console.log('onSubmit, values: ' + JSON.stringify(values));
  };

  const randomizeDefaultValues = () => {
    setDefaultValues({
      myInput: `default ${Math.floor(Math.random() * 100)}`,
    });
  };

  return (
    <Form className="demo" onSubmit={handleSubmit} defaultValues={defaultValues}>
      <Input name="myInput" label="Example input" />

      <p className="mt-4">Current default value: {defaultValues.myInput}</p>

      <button type="submit">Submit</button>
      <button type="reset">Reset</button>
      <button type="button" onClick={randomizeDefaultValues} className="ml-1">
        Update default values
      </button>
    </Form>
  );
}

<Example />;
```

#### Values

Override the values of the form fields. Changing those values will override the Field value, even if the user changed it.

```jsx
import React, { useState } from 'react';
import { Form, Input } from 'react-ocean-forms';

function Example() {
  const [values, setValues] = useState({ myInput: 'demo value' });

  // Submit callback, here you'd make your api calls
  const handleSubmit = values => {
    console.log('onSubmit, values: ' + JSON.stringify(values));
  };

  const randomizeValues = () => {
    setValues({
      myInput: `demo ${Math.floor(Math.random() * 100)}`,
    });
  };

  return (
    <Form className="demo" onSubmit={handleSubmit} values={values}>
      <Input name="myInput" label="Example input" />

      <button type="submit">Submit</button>
      <button type="reset">Reset</button>
      <button type="button" onClick={randomizeValues} className="ml-1">
        Update values
      </button>
    </Form>
  );
}

<Example />;
```

#### Disabled form

Setting the disabled prop will disable all form fields.

```jsx
import { Form, Input } from 'react-ocean-forms';

function Example() {
  // Submit callback, here you'd make your api calls
  const handleSubmit = values => {
    console.log('onSubmit, values: ' + JSON.stringify(values));
  };

  return (
    <Form className="demo" onSubmit={handleSubmit} disabled>
      <Input name="myInput" label="Example input 1" />
      <Input name="myInput2" label="Example input 2" />
      <button type="submit">Submit</button>
    </Form>
  );
}

<Example />;
```

#### Plaintext form

Setting the plaintext prop will show all form fields in a text-only mode.

```jsx
import { Form, Input } from 'react-ocean-forms';

function Example({ logMessage }) {
  const defaultValues = {
    myInput: 'default value 1',
    myInput2: 'default value 2',
  };

  return (
    <Form className="demo" defaultValues={defaultValues} plaintext>
      <Input name="myInput" label="Example input 1" />
      <Input name="myInput2" label="Example input 2" />
    </Form>
  );
}

<Example />;
```

#### Form-wide validation

Use a form-wide validation function before submit. Notice that the onSubmit callback is not invoked if you type 'bad' into the input.

```jsx
import { Form, Input, ValidationSummary } from 'react-ocean-forms';

function Example() {
  // Submit callback, here you'd make your api calls
  const handleSubmit = values => {
    console.log('onSubmit, values: ' + JSON.stringify(values));
  };

  // Reset callback, can be useful in some cases for cleanup
  const handleReset = () => {
    console.log('onReset');
  };

  // Form wide validation function
  const handleValidate = values => {
    if (values.demoInput === 'bad') {
      return {
        demoInput: 'Invalid input!',
      };
    }

    return null;
  };

  return (
    <Form
      className="demo"
      onSubmit={handleSubmit}
      onReset={handleReset}
      onValidate={handleValidate}
    >
      <ValidationSummary id="summary" />

      <Input name="demoInput" label="Example input" />

      <button type="submit">Submit</button>
      <button type="reset">Reset</button>
    </Form>
  );
}

<Example />;
```
