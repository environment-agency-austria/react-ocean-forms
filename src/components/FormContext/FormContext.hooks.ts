import { useContext } from 'react';

import { FormContext } from './FormContext';
import { IFieldValues, IFormContext } from './FormContext.types';

// tslint:disable-next-line:export-name naming-convention
export const useFormContext = <TFieldValues = IFieldValues>(): IFormContext<TFieldValues> => {
  const context = useContext(<React.Context<IFormContext<TFieldValues> | undefined>>FormContext);
  if (context === undefined) {
    throw new Error('[useFormContext]: Could not find form context. This component must be used inside a <Form> tag.');
  }

  return context;
};
