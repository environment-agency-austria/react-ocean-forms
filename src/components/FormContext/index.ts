export * from './FormContext';
export * from './FormContext.types';
export * from './FormContext.hooks';
