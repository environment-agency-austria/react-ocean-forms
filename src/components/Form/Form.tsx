/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';

import { getDeepValue, parseValidationError } from '../../utils';
import { stringFormatter as defaultStringFormatter } from '../../utils/stringFormatter';
import { FormContext, IBaseFormContext, IFieldState, IFieldValues, IFormContext, TFormEventListener } from '../FormContext';
import { IValidationState } from '../withValidation';
import { IFormProps } from './Form.types';

interface IFormState<TFieldValues = IFieldValues> {
  context: IBaseFormContext<TFieldValues>;
}

/**
 * Wrapper for managed forms
 */
export class Form<TFieldValues = IFieldValues, TSubmitArgs = unknown>
extends React.Component<IFormProps<TFieldValues, TSubmitArgs>, IFormState<TFieldValues>> {
  public static displayName: string = 'Form';

  // tslint:disable-next-line:typedef
  public static defaultProps = {
    defaultValues: {},
    asyncValidationWait: 400,
    asyncValidateOnChange: false,
    formatString: defaultStringFormatter,
    disabled: false,
    plaintext: false,
  };

  private readonly fields: Map<string, IFieldState> = new Map();
  private readonly eventListeners: Map<string, TFormEventListener> = new Map();

  constructor(props: IFormProps<TFieldValues>) {
    super(props);

    this.state = {
      context: {
        fieldPrefix: null,

        registerField: this.registerField,
        unregisterField: this.unregisterField,
        notifyFieldEvent: this.notifyFieldEvent,

        registerListener: (name: string, callback: TFormEventListener): void => { this.eventListeners.set(name, callback); },
        unregisterListener: (name: string): void => { this.eventListeners.delete(name); },

        getFieldState: this.getFieldState,
        getValues: this.getValues,

        submit: this.submit,

        busy: false,
      },
    };
  }

  /**
   * Returns the current state of the given field
   * @param name Field name
   * @returns Current field state or default field state
   */
  private getFieldState = (name: string): IFieldState => {
    const fieldState = this.fields.get(name);
    if (fieldState === undefined) {
      throw new Error(`[Form] getFieldState: Could not find state of field '${name}'`);
    }

    return fieldState;
  }

  /**
   * Generates and returns an object that contains
   * all values from all the fields.
   * @returns Current values in form of { name: value, name2: value2, ... }
   */
  private getValues = (): TFieldValues => {
    const values: IFieldValues = {};

    this.fields.forEach((state, name) => {
      if (state.isGroup === true) { return; }

      const nameParts = name.split('.');
      let valueRef = values;

      nameParts.forEach((key, index) => {
        if (nameParts.length === 1 || index === nameParts.length - 1) {
          valueRef[key] = state.getValue();
        } else {
          if (valueRef[key] === undefined) { valueRef[key] = {}; }
          valueRef = valueRef[key] as IFieldValues;
        }
      });
    });

    return (values as unknown) as TFieldValues;
  }

  /**
   * Gets called when a field triggers an event
   * @param name Field name
   * @param event Event name
   * @param args Event args
   */
  private notifyFieldEvent = (name: string, event: string, args?: unknown): void => {
    if (event === 'validation') {
      const { label } = this.getFieldState(name);
      this.notifyListeners(name, event, { ...args, label });
    } else {
      this.notifyListeners(name, event, args);
    }
  }

  /**
   * Notifies the event listeners about an event
   * @param name Field name
   * @param event Event name
   * @param args Event args
   */
  private notifyListeners(name: string, event: string, args?: unknown): void {
    this.eventListeners.forEach((callback) => {
      callback(name, event, args);
    });
  }

  /**
   * Handles the submit event of the form - prevents
   * the default and runs the submit logic
   * @param event Event object
   */
  private handleSubmit = (event: React.FormEvent): void => {
    event.preventDefault();
    event.stopPropagation();
    void this.submit();
  }

  /**
   * Submits the form - triggers any validations if needed
   * and raises the onFormSubmit prop callback if the
   * form is currently valid.
   * @param submitArgs Arguments that will be passed
   * to the onSubmit callback
   */
  private submit = async (submitArgs?: TSubmitArgs): Promise<void> => {
    this.updateBusyState(true);

    // Iterate through all fields and validate them
    // if needed.
    const validations: Promise<IValidationState>[] = [];
    this.fields.forEach(field => {
      validations.push(field.validate({
        checkAsync: true,
        immediateAsync: true,
      }));
    });

    const validationStates = await Promise.all(validations);

    // Check if all fields are valid
    const allValid = validationStates.every(state => state.valid === true);
    if (allValid === false) {
      this.notifyFieldEvent('_form', 'submit-invalid');
      this.updateBusyState(false);

      return;
    }

    // Call the form wide validation
    const formValid = this.triggerFormValidation();
    if (formValid === false) {
      this.notifyFieldEvent('_form', 'submit-invalid');
      this.updateBusyState(false);

      return;
    }

    // Await the result from callOnSubmit
    const callOnSubmitResult = this.callOnSubmit(submitArgs);
    const { resetOnSubmit } = this.props;

    // Make sure the state is cleaned up before
    const cleanup = (resetForm: boolean | undefined): void => {
      this.updateBusyState(false);
      if (resetForm) { this.reset(); }
    };
    if (callOnSubmitResult instanceof Promise) {
      void callOnSubmitResult.then(
        () => { cleanup(resetOnSubmit); },
      );
    } else {
      cleanup(resetOnSubmit);
    }
  }

  /**
   * Updates the busy state in the form context
   * @param busy Busy state
   */
  private updateBusyState(busy: boolean): void {
    this.setState(prevState => ({
      context: {
        ...prevState.context,
        busy,
      },
    }));
  }

  /**
   * Triggers the form wide validation callback if given
   * @returns Validation state of the form
   */
  private triggerFormValidation(): boolean {
    const { onValidate } = this.props;
    if (onValidate === undefined) { return true; }

    const values = this.getValues();
    const result = onValidate(values);

    // If the callback returned null then the form is valid
    if (result === null) { return true; }

    // Otherwise parse the result object and update the
    // field states.
    let allFieldsValid = true;
    this.fields.forEach((state, name) => {
      const fieldError = parseValidationError(name, getDeepValue(name, result));
      const isValid = fieldError === null || typeof fieldError !== 'object';

      if (isValid) { return; }

      state.updateValidation({
        valid: false,
        error: fieldError,
      });
      allFieldsValid = false;
    });

    return allFieldsValid;
  }

  /**
   * Gathers all the data and calls the onSubmit callback
   * if provided.
   * @param submitArgs Arguments that will be passed
   * to the onSubmit callback
   */
  private callOnSubmit(submitArgs?: TSubmitArgs): void | Promise<void> {
    const { onSubmit } = this.props;
    if (onSubmit === undefined) { return; }

    const values = this.getValues();

    return onSubmit(values, submitArgs);
  }

  /**
   * Handles the reset event of the form - prevents
   * the default and sets the state of all fields back
   * to the default state.
   * @param event Event object
   */
  private handleReset = (event: React.FormEvent): void => {
    event.preventDefault();
    event.stopPropagation();
    this.reset();
  }

  /**
   * Sets the state of all fields back
   * to the default state.
   */
  private reset = (): void => {
    this.fields.forEach((state) => {
      state.reset();
    });

    const { onReset } = this.props;
    if (onReset !== undefined) {
      onReset();
    }
  }

  /**
   * Registers a new field to the form.
   * @param name Field name
   * @param fieldState Field state
   */
  private registerField = (name: string, fieldState: IFieldState): void => {
    if (typeof name !== 'string' || name.length === 0) {
      throw new Error('[Form] registerField: name is required');
    }

    if (typeof fieldState !== 'object') {
      throw new Error('[Form] registerField: field state is required');
    }

    if (
      typeof fieldState.label !== 'string'
      || typeof fieldState.validate !== 'function'
      || typeof fieldState.updateValidation !== 'function'
      || typeof fieldState.reset !== 'function'
      || typeof fieldState.getValue !== 'function'
    ) {
      throw new Error('[Form] registerField: invalid field state given');
    }

    this.fields.set(name, fieldState);
  }

  /**
   * Unregisters a field from the form.
   * @param name Field name
   */
  private unregisterField = (name: string): void => {
    const { label } = this.getFieldState(name);
    this.notifyListeners(name, 'validation', {
      label,
      valid: true,
    });
    this.fields.delete(name);
  }

  /**
   * Combines the local form context with
   * the values from the props to form the
   * full form context passed to the form
   * components.
   */
  private prepareFormContext(): IFormContext<TFieldValues> {
    const { context } = this.state;
    const {
      defaultValues,
      values,
      asyncValidateOnChange,
      asyncValidationWait,
      formatString: stringFormatter,
      disabled,
      plaintext,
      busy: busyProp,
    } = this.props;

    // Override the busy state with the busy prop if it is set to true
    const busy = busyProp === true ? busyProp : context.busy;

    return {
      ...context,
      defaultValues,
      values,
      asyncValidationWait,
      stringFormatter,
      disabled,
      plaintext,
      busy,
      asyncValidateOnChange: asyncValidateOnChange,
    };
  }

  /**
   * Renders the form and wraps all its children
   * in a FormContext provider and a html form.
   */
  // tslint:disable-next-line:member-ordering
  public render(): JSX.Element {
    const {
      children,
      className,
      plaintext,
    } = this.props;

    const context = this.prepareFormContext();

    let formClass = className === undefined ? '' : className;
    if (plaintext) { formClass = `${formClass} plaintext`; }

    // tslint:disable-next-line:naming-convention
    const TypedFormContext = (FormContext as unknown) as React.Context<IFormContext<TFieldValues | undefined>>;

    return (
      <TypedFormContext.Provider value={context}>
        <form className={formClass} onSubmit={this.handleSubmit} onReset={this.handleReset}>
          {children}
        </form>
      </TypedFormContext.Provider>
    );
  }
}
