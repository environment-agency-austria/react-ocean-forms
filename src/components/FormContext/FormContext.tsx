/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';
import { IFormContext } from './FormContext.types';

/**
 * Context for the communication between the form
 * and its fields.
 */
export const FormContext = React.createContext<IFormContext | undefined>(undefined);
