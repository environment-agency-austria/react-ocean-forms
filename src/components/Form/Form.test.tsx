import * as React from 'react';

// tslint:disable-next-line:no-implicit-dependencies
import { shallow, ShallowWrapper } from 'enzyme';
import { mockEvent } from '../../test-utils/enzymeEventUtils';
import { IFieldState, IFormContext, TFieldValues, TFormEventListener } from '../FormContext';
import { Form } from './Form';
import { IFormProps } from './Form.types';

describe('<Form />', () => {
  interface ISetupArgs {
    props?: Partial<IFormProps>;
  }

  interface ISetupResult {
    wrapper: ShallowWrapper;
    formContext: IFormContext;
    form: ShallowWrapper;
  }

  const setup = ({
    props,
  }: Partial<ISetupArgs> = {}): ISetupResult => {
    const wrapper = shallow((
      <Form
        {...props}
      >
        <div>unitChild</div>
      </Form>
    ));

    const formContext = wrapper.first().prop('value');
    const form = wrapper.find('form');

    return {
      wrapper,
      formContext,
      form,
    };
  };

  const createMockFieldState = (label: string, isGroup?: boolean): IFieldState => ({
    label,
    isGroup,
    validate: jest.fn().mockResolvedValue({
      isValidating: false,
      valid: true,
      error: null,
    }),
    updateValidation: jest.fn(),
    reset: jest.fn(),
    getValue: jest.fn().mockReturnValue(label),
  });

  interface IMockField {
    name: string;
    state: IFieldState;
  }

  const createMockField = (name: string, label: string, isGroup?: boolean): IMockField => ({
    name,
    state: createMockFieldState(label, isGroup),
  });

  const registerUnitField = (fields: IMockField[], formContext: IFormContext): void => {
    fields.forEach(field => formContext.registerField(field.name, field.state));
  };

  interface IMockListener {
    id: string;
    state: TFormEventListener;
  }

  const createMockListener = (id?: string): IMockListener => ({
    id: id || 'listener',
    state: jest.fn(),
  });

  const createMockListeners = (count: number): IMockListener[] => {
    const result = [];
    for (let i = 0; i < count; i += 1) {
      result.push(createMockListener(`listener${i}`));
    }

    return result;
  };

  it('should render without error', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should create a valid form context', () => {
    const { formContext } = setup();
    expect(formContext).toMatchObject({
      fieldPrefix: null,
      registerField: expect.any(Function),
      unregisterField: expect.any(Function),
      notifyFieldEvent: expect.any(Function),
      registerListener: expect.any(Function),
      unregisterListener: expect.any(Function),
      getFieldState: expect.any(Function),
      getValues: expect.any(Function),
      busy: false,
      disabled: false,
      asyncValidateOnChange: false,
      asyncValidationWait: 400,
      defaultValues: {},
      values: undefined,
    });
  });

  describe('configuration', () => {
    const cases = [
      ['disabled', true, 'disabled'],
      ['formatString', jest.fn(), 'stringFormatter'],
      ['asyncValidateOnChange', true, 'asyncValidateOnChange'],
      ['asyncValidationWait', 800, 'asyncValidationWait'],
      ['defaultValues', { foo: 'bar' }, 'defaultValues'],
      ['values', { foo: 'bar' }, 'values'],
      ['plaintext', true, 'plaintext'],
    ];

    test.each(cases)('case %s', (prop, value, contextProp) => {
      const { formContext } = setup({
        props: { [prop]: value },
      });
      expect(formContext).toMatchObject({ [contextProp]: value });
    });
  });

  describe('invalid field registration', () => {
    // tslint:disable-next-line:no-empty
    const mf = (): void => {};
    const cases = [
      ['no parameters', undefined, undefined],
      ['invalid field name', '', undefined],
      ['no state', 'foo', undefined],
      ['empty state', 'foo', {}],
      ['1 of 5 props', 'foo', { label: 'hey' }],
      ['2 of 5 props', 'foo', { label: 'hey', validate: mf }],
      ['3 of 5 props', 'foo', { label: 'hey', validate: mf, updateValidation: mf }],
      ['4 of 5 props', 'foo', {
        label: 'hey',
        validate: mf,
        updateValidation: mf,
        reset: mf,
      }],
    ];

    test.each(cases)('case %s', (testName, fieldName, fieldState) => {
      const { formContext } = setup();
      expect(() => formContext.registerField(fieldName, fieldState)).toThrowErrorMatchingSnapshot();
    });
  });

  describe('field states and values', () => {
    const createCases = (): (string | IMockField)[][] => {
      return [
        ['field', createMockField('unitField', 'Unit field')],
        ['group', createMockField('unitGroup', 'Unit group', true)],
        ['sub field', createMockField('unitGroup.subField', 'Sub field', true)],
      ];
    };

    describe('formContext.registerField - field registration', () => {
      const cases = createCases();

      test.each(cases)('should register a new %s without crashing', (testName, field) => {
        const { formContext } = setup();
        expect(() => registerUnitField([field], formContext)).not.toThrowError();
      });
    });

    describe('formContext.unregisterField - field cleanup', () => {
      const cases = createCases();

      test.each(cases)('should unregister %s without crashing', (testName, field) => {
        const { formContext } = setup();
        formContext.registerField(field.name, field.state);
        formContext.unregisterField(field.name);
      });
    });

    describe('formContext.getFieldState - field states', () => {
      const cases = createCases();

      test.each(cases)('should return the correct field state of a %s', (testName, field) => {
        const { formContext } = setup();
        registerUnitField([field], formContext);
        expect(formContext.getFieldState(field.name)).toBe(field.state);
      });
    });

    describe('formContext.getValues - form values', () => {
      const { formContext } = setup();

      const unitField = createMockField('unitField', 'Unit field');
      const unitGroup = createMockField('unitGroup', 'Unit group', true);
      const unitSubField = createMockField(`${unitGroup.name}.subField`, 'Sub field');

      const mockFields = [unitField, unitGroup, unitSubField];
      registerUnitField(mockFields, formContext);

      let formValues: TFieldValues;
      it('should return the values without crashing', () => {
        expect(() => {
          formValues = formContext.getValues();
        }).not.toThrowError();
      });

      it('should call unitField.getValue', () => {
        expect(unitField.state.getValue).toHaveBeenCalled();
      });

      it('should call subField.getValue', () => {
        expect(unitSubField.state.getValue).toHaveBeenCalled();
      });

      it('should not call group.getValue', () => {
        // The value of fieldGroups is computed by its children
        expect(unitGroup.state.getValue).not.toHaveBeenCalled();
      });

      it('should return the correct form values', () => {
        const subFieldLocalName = unitSubField.name.substring(unitGroup.name.length + 1);
        const expectedFormValues = {
          [unitField.name]: unitField.state.label,
          [unitGroup.name]: {
            [subFieldLocalName]: unitSubField.state.label,
          },
        };

        expect(formValues).toMatchObject(expectedFormValues);
      });
    });
  });

  describe('listener / notify system', () => {
    interface ISetupListenerResult extends ISetupResult {
      unitField: IMockField;
      mockListeners: IMockListener[];
    }

    const setupListener = (count: number): ISetupListenerResult => {
      const setupResult = setup();

      const unitField = createMockField('unitField', 'Unit field');
      setupResult.formContext.registerField(unitField.name, unitField.state);

      const mockListeners = createMockListeners(count);
      mockListeners.forEach(item => setupResult.formContext.registerListener(item.id, item.state));

      return {
        ...setupResult,
        mockListeners,
        unitField,
      };
    };

    it('should register new listeners without crashing', () => {
      const mockListeners = createMockListeners(3);
      const { formContext } = setup();
      mockListeners.forEach((item) => {
        expect(() => formContext.registerListener(item.id, item.state)).not.toThrowError();
      });
    });

    it('should unregister new listeners without crashing', () => {
      const mockListeners = createMockListeners(3);
      const { formContext } = setup();
      mockListeners.forEach((item) => {
        formContext.registerListener(item.id, item.state);
        expect(() => formContext.unregisterListener(item.id)).not.toThrowError();
      });
    });

    it('should pass the validation notification to all listeners', () => {
      const eventName = 'validation';
      const eventArgs = { foo: 'bar' };

      const { formContext, unitField, mockListeners } = setupListener(3);

      formContext.notifyFieldEvent(unitField.name, eventName, eventArgs);
      mockListeners.forEach(item => expect(item.state).toHaveBeenLastCalledWith(
        unitField.name,
        eventName,
        {
          label: unitField.state.label,
          ...eventArgs,
        },
      ));
    });

    it('should call the listeners', () => {
      const eventName = 'change';
      const eventArgs = 'myNewValue';

      const { formContext, unitField, mockListeners } = setupListener(3);

      formContext.notifyFieldEvent(unitField.name, eventName, eventArgs);
      mockListeners.forEach(item => expect(item.state).toHaveBeenLastCalledWith(
        unitField.name,
        eventName,
        eventArgs,
      ));
    });
  });

  describe('onSubmit handling', () => {
    // @ts-ignore Form submit is private
    const simulateSubmitEvent = async (wrapper: ShallowWrapper): Promise<void> => (wrapper.instance() as Form).submit();

    interface ISetupSubmitArgs extends ISetupArgs {
      addListeners: boolean;
      customField: IMockField;
    }

    interface ISetupSubmitResult extends ISetupResult {
      expectedFormValues: TFieldValues;
      mockFields: IMockField[];
      mockListeners?: IMockListener[];
    }

    const unitFieldName = 'unitField';

    const setupSubmit = async ({
      props,
      customField,
      addListeners = false,
    }: Partial<ISetupSubmitArgs> = {}): Promise<ISetupSubmitResult> => {
      const result = setup({ props });

      const unitField = createMockField(unitFieldName, 'Unit field');
      const unitGroup = createMockField('unitGroup', 'Unit group', true);
      const unitSubField = createMockField(`${unitGroup.name}.subField`, 'Sub field');

      const mockFields = [unitField, unitGroup, unitSubField];
      if (customField) {
        mockFields.push(customField);
      }

      registerUnitField(mockFields, result.formContext);

      let mockListeners;

      if (addListeners) {
        mockListeners = createMockListeners(3);
        mockListeners.forEach(item => result.formContext.registerListener(item.id, item.state));
      }

      await simulateSubmitEvent(result.wrapper);

      const subFieldLocalName = unitSubField.name.substring(unitGroup.name.length + 1);
      const expectedFormValues = {
        [unitField.name]: unitField.state.label,
        [unitGroup.name]: {
          [subFieldLocalName]: unitSubField.state.label,
        },
      };

      return {
        ...result,
        expectedFormValues,
        mockFields,
        mockListeners,
      };
    };

    describe('all valid', () => {
      it('should call all the validation functions', async () => {
        const { mockFields } = await setupSubmit();

        mockFields.forEach(item => expect(item.state.validate).toHaveBeenLastCalledWith({
          checkAsync: true,
          immediateAsync: true,
        }));
      });

      it('should call the onValidate prop', async () => {
        const onValidateHandler = jest.fn().mockReturnValue(null);
        const { expectedFormValues } = await setupSubmit({ props: { onValidate: onValidateHandler }});

        expect(onValidateHandler).toHaveBeenCalledWith(expectedFormValues);
      });

      it('should call the onSubmit prop', async () => {
        const onSubmitHandler = jest.fn();
        const { expectedFormValues } = await setupSubmit({ props: { onSubmit: onSubmitHandler }});

        expect(onSubmitHandler).toHaveBeenCalledWith(expectedFormValues, undefined);
      });
    });

    describe('invalid through form validator', () => {
      const createInvalidValidator = (): jest.Mock => jest.fn().mockReturnValue({
        [unitFieldName]: 'error',
      });

      it('should update the validation state of the field', async () => {
        const { mockFields } = await setupSubmit({ props: { onValidate: createInvalidValidator() }});

        // tslint:disable-next-line:no-non-null-assertion
        expect(mockFields.find(item => item.name === unitFieldName)!.state.updateValidation).toHaveBeenCalledWith({
          valid: false,
          error: {
            message_id: 'error',
            params: { },
          },
        });
      });

      it('should trigger a submit-invalid event', async () => {
        const { mockListeners } = await setupSubmit({ props: { onValidate: createInvalidValidator() }, addListeners: true});
        // tslint:disable-next-line:no-non-null-assertion
        mockListeners!.forEach(item => expect(item.state).toHaveBeenLastCalledWith(
          '_form',
          'submit-invalid',
          undefined,
        ));
      });

      it('should not call the onSubmit prop', async () => {
        const onSubmitHandler = jest.fn();
        await setupSubmit({ props: {
          onValidate: createInvalidValidator(),
          onSubmit: onSubmitHandler,
        }});
        expect(onSubmitHandler).not.toHaveBeenCalled();
      });
    });

    describe('invalid through field validators', () => {
      const createInvalidField = (): IMockField => {
        const field = createMockField('invalidField', 'Invalid field');
        field.state.validate = jest.fn().mockResolvedValue({
          isValidating: false,
          valid: false,
          error: 'foobar',
        });

        return field;
      };

      it('should trigger a submit-invalid event', async () => {
        const { mockListeners } = await setupSubmit({ customField: createInvalidField(), addListeners: true});

        // tslint:disable-next-line:no-non-null-assertion
        mockListeners!.forEach(item => expect(item.state).toHaveBeenLastCalledWith(
          '_form',
          'submit-invalid',
          undefined,
        ));
      });

      it('should not call the onSubmit prop', async () => {
        const onSubmitHandler = jest.fn();
        await setupSubmit({
          props: { onSubmit: onSubmitHandler },
          customField: createInvalidField(),
        });
        expect(onSubmitHandler).not.toHaveBeenCalled();
      });
    });

    describe('context busy state', () => {
      const testBusyState = (wrapper: ShallowWrapper, formContext: IFormContext, expected: boolean, done: jest.DoneCallback): void => {
        simulateSubmitEvent(wrapper);

        process.nextTick(() => {
          wrapper.update();
          expect(formContext.busy).toBe(expected);
          done();
        });
      };

      it('should not be busy if the onSubmit callback returns immediately', (done) => {
        const onSubmitHandler = jest.fn();
        const { wrapper, formContext } = setup({ props: { onSubmit: onSubmitHandler }});
        testBusyState(wrapper, formContext, false, done);
      });

      it('should not be busy if there is no onSubmit callback', (done) => {
        const { wrapper, formContext } = setup();
        testBusyState(wrapper, formContext, false, done);
      });

      describe('async onSubmit callback', () => {
        const createSlowOnSubmit = (): () => Promise<void> => {
          return (): Promise<void> => new Promise(
            (resolve: Function): NodeJS.Timer => setTimeout(
              (): void => { resolve(); },
              1000,
            ),
          );
        };

        beforeAll(() => {
          jest.useFakeTimers();
        });

        afterAll(() => {
          jest.useRealTimers();
        });

        it('should be busy after invoking onSubmit', async (done) => {
          const { wrapper } = setup({ props: { onSubmit: createSlowOnSubmit() }});
          await simulateSubmitEvent(wrapper);

          process.nextTick(() => {
            wrapper.update();
            const formContext: IFormContext = wrapper.first().prop('value');
            expect(formContext.busy).toBe(true);
            done();
          });
        });

        it('should not be busy after onSubmit finished', async (done) => {
          const { wrapper } = setup({ props: { onSubmit: createSlowOnSubmit() }});
          await simulateSubmitEvent(wrapper);

          jest.runAllTimers();

          process.nextTick(() => {
            wrapper.update();
            const formContext: IFormContext = wrapper.first().prop('value');
            expect(formContext.busy).toBe(false);
            done();
          });
        });
      });
    });
  });

  describe('onReset handling', () => {
    const simulateResetEvent = (form: ShallowWrapper): ShallowWrapper => form.simulate('reset', mockEvent());

    it('should reset all fields', () => {
      const { form, formContext } = setup();

      const unitField = createMockField('unitField', 'Unit field');
      const unitGroup = createMockField('unitGroup', 'Unit group', true);
      const unitSubField = createMockField(`${unitGroup.name}.subField`, 'Sub field');
      const mockFields = [unitField, unitGroup, unitSubField];
      registerUnitField(mockFields, formContext);

      simulateResetEvent(form);
      mockFields.forEach(item => expect(item.state.reset).toHaveBeenCalled());
    });

    it('should call the onReset prop', () => {
      const onResetHandler = jest.fn();
      const { form } = setup({ props: { onReset: onResetHandler }});
      simulateResetEvent(form);

      expect(onResetHandler).toHaveBeenCalled();
    });
  });
});
