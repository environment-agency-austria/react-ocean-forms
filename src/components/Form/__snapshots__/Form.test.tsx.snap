// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> invalid field registration case 1 of 5 props 1`] = `"[Form] registerField: invalid field state given"`;

exports[`<Form /> invalid field registration case 2 of 5 props 1`] = `"[Form] registerField: invalid field state given"`;

exports[`<Form /> invalid field registration case 3 of 5 props 1`] = `"[Form] registerField: invalid field state given"`;

exports[`<Form /> invalid field registration case 4 of 5 props 1`] = `"[Form] registerField: invalid field state given"`;

exports[`<Form /> invalid field registration case empty state 1`] = `"[Form] registerField: invalid field state given"`;

exports[`<Form /> invalid field registration case invalid field name 1`] = `"[Form] registerField: name is required"`;

exports[`<Form /> invalid field registration case no parameters 1`] = `"[Form] registerField: name is required"`;

exports[`<Form /> invalid field registration case no state 1`] = `"[Form] registerField: field state is required"`;

exports[`<Form /> should render without error 1`] = `
<ContextProvider
  value={
    Object {
      "asyncValidateOnChange": false,
      "asyncValidationWait": 400,
      "busy": false,
      "defaultValues": Object {},
      "disabled": false,
      "fieldPrefix": null,
      "getFieldState": [Function],
      "getValues": [Function],
      "notifyFieldEvent": [Function],
      "plaintext": false,
      "registerField": [Function],
      "registerListener": [Function],
      "stringFormatter": [Function],
      "submit": [Function],
      "unregisterField": [Function],
      "unregisterListener": [Function],
      "values": undefined,
    }
  }
>
  <form
    className=""
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div>
      unitChild
    </div>
  </form>
</ContextProvider>
`;
